[client] Add client metadata header on RestClient requests

swallez opened the pull request:
Adds a `X-Elastic-Client-Meta` header to http requests sent by `RestClient`. This header contains information about the runtime environment that is meant to allow analyzing usage context by collecting this information on the receiving side of requests, like a proxy server in front of ES.

The information in this header has been specified by the clients team to be consistent across language clients. It contains the versions of the client, the language runtime (Java in this case) and its http library, and the use of the high level client. Additional JVM specific information is added about standard libraries for other JVM languages like Kotlin, Scala, Groovy and Clojure, so that we can extrapolate interest for dedicated clients for these languages.

This header is sent by default and can be disabled using `RestClientBuilder.setMetadataHeaderEnabled(false)`.

This PR also sets the `User-Agent` header to `elasticsearch-java/<version> (Java/<version>)` to again be consistent with other language clients. Users are however free to override this value with application-specific agent information.

swallez committed:
[client] Add client metadata header on RestClient requests

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

swallez commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch '7.x' into client-meta-header/7.x

andreidan commented on client/rest/src/main/java/org/elasticsearch/client/RuntimeInfo.java:
```diff
@@ -0,0 +1,145 @@
+     */
+    public static String getRuntimeMetadata() {
+        StringBuilder s = new StringBuilder();
+        String v;
```
Does `v` stand for `version`? If so, shall we name it that?

swallez replied:
Done!

andreidan commented on client/rest/src/main/java/org/elasticsearch/client/RuntimeInfo.java:
```diff
@@ -0,0 +1,145 @@
+            String version = (String) m.invoke(null);
+            return stripPatchRevision(version);
+
+        } catch (Throwable t) {
```
I believe we should only catch `Exception` here to allow for `OOM` errors to propagate

swallez replied:
Good point. I wanted to avoid any kind of class-loading problem to cause returning `null` but the non-exception we can encounter (besides OOM) are things like `LinkageError` that we'd better let bubble up as well.

Changed to `Exception`.

andreidan commented on client/rest/src/main/java/org/elasticsearch/client/RuntimeInfo.java:
```diff
@@ -0,0 +1,145 @@
+        return null;
+    }
+
+    private static String stripPatchRevision(String version) {
```
Can we feed this method a large input of version formats in a test to make sure it does the right thing? (I guess it'd be more about looking into the various languages versioning strategies)

swallez replied:
They all use semantic versioning but have slight differences in how they name non-final releases. I added `RuntimeInfoTests` that checks various version naming strategies for the languages we test (information found on https://search.maven.org/)

andreidan replied:
This is great, thank you!

andreidan reviewed (commented):
Thanks for working on this @swallez 

I left a few minor comments. I'll ask the @es/core-infra team about possible strategies to avoid having to keep in sync a new `VERSION` field (possibly copy/make it available via gradle?)

swallez committed:
Change Throwable to Exception, add test

swallez committed:
Rename 'v' to 'version'

swallez commented:
Since adding Kotlin/Scala/Clojure/Groovy as dependencies of LLRC tests would probably be too heavy, I've duplicated version extraction in https://github.com/swallez/jvm-languages-versions with actual dependencies to test proper version identification.

swallez commented:
@elasticmachine retest this please

rjernst commented on client/rest/src/main/java/org/elasticsearch/client/RestClient.java:
```diff
@@ -104,6 +104,11 @@
+    /** The version of the Rest client*/
+    // Must be kept in sync with org.elasticsearch.Version.CURRENT, but we can't call that class as the low level
+    // client should not depend on the server code. A test in `VersionTests` will fail if they're not equal.
+    public static final String VERSION = "7.11.0";
```
Please do not hardcode this version. Even with a test, it would add another place that needs to be updated on every version bump, increasing the complexity of the bump.  Instead, we can use the manifest attribute for Elasticsearch version that every jar contains. And we should fail if this attribute does not exist, to ensure if a user builds a fat jar with the client, we will not silently be missing the version.

swallez replied:
This is some telemetry information that users should not be bothered with. They can opt out if they want, but it should not bring additional requirements on their build system. So we'd rather fail silently and report a missing version, hoping that it will not be statistically significant.

And if it is significant (which is why reporting missing information is important too) then we can work on updating the build system so that this information exists as generated code that will survive a fat jar operation.

This is targeted at 7.x is because FF for 7.11.0 is really close and I wanted to avoid the additional time needed for a backport. So 7.x first, and backport to master even if I know this isn't the common way of doing it.

rjernst reviewed (changes requested):
One overall question: why is this targeted at 7.x? It seems like, as with other PRs, this should be going through the master branch first?

swallez committed:
Read version from Manifest to avoid hardcoding the version in the client

swallez commented:
@rjernst I added a commit that reads the version from the manifest (with gracefull fallback), and removes the hardcoded version and consequently reverted the changes to `org.elasticsearch.Version`.

swallez committed:
Fix style issue & forbidden API

rjernst commented:
> This is targeted at 7.x is because FF for 7.11.0 is really close and I wanted to avoid the additional time needed for a backport. So 7.x first, and backport to master even if I know this isn't the common way of doing it.

I don't think we should change our standard git workflow because of urgency. We go through master for many reasons. For example, it keeps the git log consistent. We also rely on the 7.x commit following the master commit temporally for BWC tests.

By making the version pulled from the manifest, I believe this PR should be identical across the branches.

rjernst commented on client/rest/src/main/java/org/elasticsearch/client/RuntimeInfo.java:
```diff
@@ -0,0 +1,145 @@
+            s.append(",sc=").append(version);
+        }
+
+        version = groovyVersion();
```
this should be `clojureVersion()`?

rjernst commented on client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java:
```diff
@@ -220,18 +264,46 @@ private CloseableHttpAsyncClient createHttpClient() {
+                HttpAsyncClientBuilder.class.getClassLoader());
+
+            // service, language, transport, followed by additional information
+            return "es=" + VERSION +
```
These are all things that will not change for the lifetime of the jvm, yet http client version and the "runtime metadata" seem to be inspected on every invocation of this method.

Instead, we could have a single method that gets invoked to generate this data, and statically cache the result. We can then be more stringent in failures, since if metadata is enabled, we shouldn't have silent failures, which in that case may only cause confusion. The error handling can be simpler then, with a single try/catch that rethrows as a runtime exception.

rjernst reviewed (changes requested):
I left a few comments

swallez commented:
@rjernst I opened PR #66303 with the same commits as this one but targetting `master` and will therefore close this one.

I also added commits that compute header values statically. They were computed at client creation time, which should happen a limited number of times, but these are indeed values that never change so it's even better to compute them statically.

Also fixed `clojureVersion()`.
